knowledge-base/adr/0019-openapi-generator-naming.md

```diff
@@ -0,0 +1,86 @@
+# Naming of generated APIs, operations, properties and query parameters (OpenAPI generator)
+
+## Status
+
+decided
+
+## Context
+
+The OpenAPI generator has to generate names for APIs, their operations and schemas.
+We initially used the file name to generate an API name, but found that for bigger APIs this is would result in very complex APIs, therefore we need a more sensible default.
+In addition API owners whish to be able to customize names of APIs and operations.
+To follow naming conventions in JavaScript/TypeScript we planned on formatting API names as pascal case, operation names as camel case and property and parameter names as camel case.
+For properties and parameters we would have to resolve the original names to use them in the requests.
+We found that for most cases this is straight forward, but there are corner cases that require a different/more complex architecture.
+
+### Issues with properties and parameters
+
+Given the following two types and their original properties, if those were used in combination, e. g. `allOf`, there is no possibility to set both values.
+
+```ts
+type A {
+  someProperty: string; // original name: 'some_property`
+}
+
+type A {
+  someProperty: string; // original name: 'some-property`
+}
+```
+
+This could be solved by making names unique throughout the whole service.
+However this again would lead to less convenient names, e.g. `name`, `name1`, `name2`.
+
+## Decisions
+
+### API naming
+
+APIs are named based on the first tag, if there are any.
+If there is no tag, a "default" tag is used.
+To allow customizing of APIs, we introduced the `x-sap-cloud-sdk-api-name` extension, which takes precedence over the default name generation.
+This extension can be set on on multiple levels:
+
+- operation, causing this operation to be treated as part of the given API.
+- path, causing all operations under this path to be treated as part of the given API.
+- document root, causing all operations in the document to be treated as part of the given API.
+
+All names are transformed to pascal case (`SomeApi`).
+API names are formed by removing "api" suffix (case independant and only if available) and adding "Api".
+Examples:
+
+- `my-api` => `MyApi`
+- `my` => `MyApi`
+- `my-api-api` => `MyApiApi`
+
+### Operation naming
+
+Operations are named based on their given `operationId`.
+If no `operationId` is given the name is parsed from the method and the path under which the operation occured.
```

Comment:
Could you perhaps given an example here of path, method etc and what will be the name generated from that.

Reply:
```suggestion
If no `operationId` is given the name is parsed from the method and the path under which the operation occured.
Example: 

- path: `entity/{id}`
- method: `get`
- resulting name: `getEntityById`
```